radio/src/opentx_helpers.h

```diff
@@ -56,13 +56,49 @@ inline void SWAP(T & a, T & b)
 
 inline void memclear(void * p, size_t size)
 {
-memset(p, 0, size);
+  memset(p, 0, size);
 }
 
 inline bool is_memclear(void * p, size_t size)
 {
-uint8_t * buf = (uint8_t *)p;
-return buf[0] == 0 && memcmp(buf, buf + 1, size - 1) == 0;
+  uint8_t * buf = (uint8_t *)p;
+  return buf[0] == 0 && memcmp(buf, buf + 1, size - 1) == 0;
 }
 
+// Circular buffer to send data between two threads
+template<class T, uint8_t N>
+class CircularBuffer {
+  protected:
+    volatile T buffer[N] = {0};
+    uint8_t readPos = 0;
+    uint8_t writePos = 0;
+    
+  public:
+    void clear()
+    {
+      memclear(this, sizeof(*this));
+    }
+
+    T read()
+    {
+      T value = buffer[readPos];
+      if (value) {
+        buffer[readPos] = 0;
+        readPos = (readPos + 1) % N;
+      }
+      return value;
+    }
+
+    // Values must be non-zero
+    bool write(T value)
+    {
+      if (value && !buffer[writePos]) {
+        buffer[writePos] = value;
+        writePos = (writePos + 1) % N;
+        return false;
+      }    
+      return true;
+    }  
```

Comment:
Do I understand correctly that to prevent read- and writepointers (readPos and writePos) overrunning each-other, it is assumed by CicularBuffer, that a value of 0 means an empty/writable cell? If so, then this class cannot be used to save value 0 (possibly not necessary in this particular application/case, but it is not a generic CB then).

Reply:
Yes, that is correct!

Reply:
To elaborate a little further - by using the same byte for both data and signal, I can get away with using `volatile`s. But if it were to be a true CB that could take any value including zero, then I would need to use e.g. the current write and read positions for synchronization. And that would require that the reader reads data *before* updating the "official" read position. That in turn would require fences like e.g. std::atomic_thread_fence. I don't know if we even have that available, but it would be more complicated for sure! The next step up the ladder would be FreeRTOS mutexes.
So I am hoping that this can pass review and testing, because it would so simple and not resource demanding!